server/src/main/java/org/elasticsearch/action/index/IndexAction.java

```diff
@@ -9,13 +9,14 @@
 package org.elasticsearch.action.index;
 
 import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.DocWriteResponse;
 
-public class IndexAction extends ActionType<IndexResponse> {
+public class IndexAction extends ActionType<DocWriteResponse> {
 
     public static final IndexAction INSTANCE = new IndexAction();
     public static final String NAME = "indices:data/write/index";
 
     private IndexAction() {
-        super(NAME, IndexResponse::new);
+        super(NAME, in -> { throw new UnsupportedOperationException(); });
```

Comment:
Just to be safe we can preserve today's behaviour in prod whilst still failing tests if anyone calls this in future:

```suggestion
        super(NAME, in -> {
            assert false : "might not be an IndexResponse!";
            return new IndexResponse(in);
        });
```